src/views/Nft/market/hooks/useCollectionNfts.tsx

```diff
@@ -0,0 +1,268 @@
+    }
+  }
+
+  if (tokenIds.length) {
+    const nftsMarket = await getMarketDataForTokenIds(collection.address, tokenIds)
+
+    const responsesPromises = tokenIds.map(async (id) => {
+      const apiMetadata: ApiSingleTokenData = collectionNftsResponse
+        ? collectionNftsResponse.data[id]
+        : await getNftApi(collection.address, id)
+      if (apiMetadata) {
+        const marketData = nftsMarket.find((nft) => nft.tokenId === id)
+
+        return {
+          tokenId: id,
+          name: apiMetadata.name,
+          description: apiMetadata.description,
+          collectionName: apiMetadata.collection.name,
+          collectionAddress: collection.address,
+          image: apiMetadata.image,
+          attributes: apiMetadata.attributes,
+          marketData,
+        }
+      }
+      return null
+    })
+
+    const responseNfts: NftToken[] = (await Promise.all(responsesPromises)).filter((x) => x)
+    newNfts.push(...responseNfts)
+    return newNfts
+  }
+  return []
+}
+
+export const useCollectionNfts = (collectionAddress: string) => {
+  const fetchedNfts = useRef<NftToken[]>([])
+  const fallbackMode = useRef(false)
+  const fallbackModePage = useRef(0)
+  const isLastPage = useRef(false)
+  const collection = useGetCollection(collectionAddress)
+  const { field, direction } = useGetNftOrdering(collectionAddress)
+  const showOnlyNftsOnSale = useGetNftShowOnlyOnSale(collectionAddress)
+  const nftFilters = useGetNftFilters(collectionAddress)
+  const [itemListingSettings, setItemListingSettings] = useState<ItemListingSettings>({
+    field,
+    direction,
+    showOnlyNftsOnSale,
+    nftFilters,
+  })
+
+  // We don't know the amount in advance if nft filters exist
+  const resultSize = !Object.keys(nftFilters).length
+    ? showOnlyNftsOnSale || field !== 'tokenId'
+      ? collection.numberTokensListed
+      : collection.totalSupply
+    : null
+
+  const itemListingSettingsJson = JSON.stringify(itemListingSettings)
+  const filtersJson = JSON.stringify(nftFilters)
+
+  useEffect(() => {
+    setItemListingSettings(() => ({
+      field,
+      direction,
+      showOnlyNftsOnSale,
+      nftFilters: JSON.parse(filtersJson),
+    }))
+    fallbackMode.current = false
+    fallbackModePage.current = 0
+    fetchedNfts.current = []
+    isLastPage.current = false
+  }, [field, direction, showOnlyNftsOnSale, filtersJson])
+
+  const {
+    data: nfts,
+    status,
+    size,
+    setSize,
+  } = useSWRInfinite(
+    (pageIndex, previousPageData) => {
+      if (pageIndex !== 0 && previousPageData && !previousPageData.length) return null
+      return [collectionAddress, itemListingSettingsJson, pageIndex, 'collectionNfts']
+    },
+    async (address, settingsJson, page) => {
+      const settings: ItemListingSettings = JSON.parse(settingsJson)
+      const tokenIdsFromFilter = await fetchTokenIdsFromFilter(collection.address, settings)
```

Comment:
`collection` might be `undefined`. Should add a gaurd here `collection?.address` ?

Reply:
Done